README.md
[![GPLv3 License](https://img.shields.io/badge/License-GPL%20v3-yellow.svg)](https://opensource.org/licenses/)
![CRAN checks](https://cranchecks.info/badges/summary/TreeLS)
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/tiagodc/TreeLS)
![](https://cranlogs.r-pkg.org/badges/grand-total/TreeLS)

# TreeLS

High performance R functions for forest data processing based on **T**errestrial **L**aser **S**canning (but not only) point clouds.

## Description

This package is a refactor of the methods described in [this paper](https://doi.org/10.1016/j.compag.2017.10.019), among many other features for 3D point cloud processing of forest environments.

Most algorithms are written in C++ and wrapped in R functions through `Rcpp`. *TreeLS* is built on top of [lidR](https://github.com/Jean-Romain/lidR/), using its `LAS` infrastructure internally for most methods.

For any questions, comments or bug reports please submit an [issue](https://github.com/tiagodc/TreeLS/issues) here on GitHub. Suggestions, ideas and references of new algorithms are always welcome - as long as they fit into TreeLS' scope.

`TreeLS` is currently on v2.0. To install it from an official mirror, use: `install.packages("TreeLS")`. To install the most recent version, check out the *Installation from source* section below.

*TreeLS is not on CRAN at the moment (August/2020), the up-to-date version is submitted and should be available shortly. Meanwhile you can install it from source using devtools.

## News

- August/2020: Version 2.0 is finally available! It's a major release, introducing several new functionalities, bug fixes, more robust estimators for noisy clouds and more flexible plotting. All functionalities from older versions are now available and optimized, so there should be no need to use legacy code anymore. The scope of application of TreeLS has become much wider in this version, specially due to the introduction of functions like `fastPointMetrics` and `shapeFit`, making it much easier for researchers to assess point cloud data in many contexts and develop their own methods on top of those functions. For a comprehensive list of the updates check out the [CHANGELOG](https://github.com/tiagodc/TreeLS/blob/master/CHANGELOG.md).

- March/2019: `TreeLS` is finally available on CRAN and is now an official R package.

<img align="right" height="400" src="https://raw.githubusercontent.com/tiagodc/Scripts/master/animations/treedt.gif">

## Main functionalities

- Tree detection at plot level
- Tree region assignment
- Stem detection and denoising
- Stem segmentation
- Forest inventory
- Fast calculation of point features
- Research basis and other applications
- 3D plotting and manipulation

## Installation from source

### Requirements
- `devtools`: run `install.packages('devtools', dependencies = TRUE)` from the R console
- Rcpp compiler:
    - on Windows: install [Rtools](https://cran.r-project.org/bin/windows/Rtools/) for your R version - make sure to add it to your system's *path*
    - on Mac: install Xcode
    - on Linux: be sure to have `r-base-dev` installed

### Install TreeLS latest version

On the R console, run:
```
devtools::install_github('tiagodc/TreeLS')
```

## Usage

Example of full processing workflow from reading a point cloud file until stem segmentation of a forest plot:
```
library(TreeLS)

# open sample plot file
file = system.file("extdata", "pine_plot.laz", package="TreeLS")
tls = readTLS(file)

# normalize the point cloud
tls = tlsNormalize(tls, keep_ground = F)
x = plot(tls)

# extract the tree map from a thinned point cloud
thin = tlsSample(tls, smp.voxelize(0.02))
map = treeMap(thin, map.hough(min_density = 0.1), 0)
add_treeMap(x, map, color='yellow', size=2)

# classify tree regions
tls = treePoints(tls, map, trp.crop())
add_treePoints(x, tls, size=4)
add_treeIDs(x, tls, cex = 2, col='yellow')

# classify stem points
tls = stemPoints(tls, stm.hough())
add_stemPoints(x, tls, color='red', size=8)

# make the plot's inventory
inv = tlsInventory(tls, d_method=shapeFit(shape='circle', algorithm = 'irls'))
add_tlsInventory(x, inv)

# extract stem measures
seg = stemSegmentation(tls, sgt.ransac.circle(n = 20))
add_stemSegments(x, seg, color='white', fast=T)

# plot everything once
tlsPlot(tls, map, inv, seg, fast=T)

# check out only one tree
tlsPlot(tls, inv, seg, tree_id = 11)

#------------------------------------------#
### overview of some new methods on v2.0 ###
#------------------------------------------#

file = system.file("extdata", "pine.laz", package="TreeLS")
tls = readTLS(file) %>% tlsNormalize()

# calculate some point metrics
tls = fastPointMetrics(tls, ptm.knn())
x = plot(tls, color='Verticality')

# get its stem points
tls = stemPoints(tls, stm.eigen.knn(voxel_spacing = .02))
add_stemPoints(x, tls, size=3, color='red')

# get dbh and height
dbh_algo = shapeFit(shape='cylinder', algorithm = 'bf', n=15, inliers=.95, z_dev=10)
inv = tlsInventory(tls, hp = .95, d_method = dbh_algo)
add_tlsInventory(x, inv)

# segment the stem usind 3D cylinders and getting their directions
seg = stemSegmentation(tls, sgt.irls.cylinder(n=300))
add_stemSegments(x, seg, color='blue')

# check out a specific tree segment
tlsPlot(seg, tls, segment = 3)

```


CHANGELOG.md
#### v2.0.0
- *ply* files are now supported in `readTLS`.
- naming convention changed for all methods used inside other functions - all start with a 3 letter prefix (`ptm`, `map`, `ptm`, `trp`...).
- some methods were renamed: `tlsTransform`, `treeMap.positions`.
- new methods added por tree mapping, stem points and stem segmentation, contemplating several new possible use cases and forest scenarios (complex forests, high accuracy LiDAR sensors, irregular stem shapes and so on).
- `treeMap.merge` function introduced and applied internally in `treeMap`, allowing detection and merging of duplicates, mapping errors, and forked stems into the same tree. 
- extra step added to the inventory workflow and new method added: `treePoints`. This way whole trees are assigned TreeIDs in a plot, not just their stem points.
- new method added for quick forest inventory metrics: `tlsInventory`.
- new functionalities introduced: `fastPointMetrics`, `shapeFit`, `nnFilter`, `writeTLS`, `shapeFit.forks`
- all RANSAC methods have been tuned internally for extra robustness and stability when estimating diameters of noisy point clouds.
- `tlsPlot` completely refactored to dynamically interpret any unnamed inputs and plot them accordingly. 
- `add_*` plot functions added, making plotting of 3D products much more flexible and compatible with `lidR` plots.
- actual 3D cylinder plotting added for all stem segmentation methods.
- dependency issues fixed and updated.
- `lidR` internal calls updated with new naming convention introduced in v3.0.0.
- bug fixes and performance enhancements.

#### v1.0.1
- bug fix #3.

R/map_methods.R
# ===============================================================================
#
# Developers:
#
# Tiago de Conto - tdc.florestal@gmail.com -  https://github.com/tiagodc/
#
# COPYRIGHT: Tiago de Conto, 2020
#
# This piece of software is open and free to use, redistribution and modifications
# should be done in accordance to the GNU General Public License >= 3
# Use this software as you wish, but no warranty is provided whatsoever. For any
# comments or questions on TreeLS, please contact the developer (prefereably through my github account)
#
# If publishing any work/study/research that used the tools in TreeLS,
# please don't forget to cite the proper sources!
#
# Enjoy!
#
# ===============================================================================

#' Tree mapping algorithm: Hough Transform
#' @description This function is meant to be used inside \code{\link{treeMap}}. It applies an adapted version of the Hough Transform for circle search. Mode details are given in the sections below.
#' @template param-min_h-max_h
#' @template param-h_step
#' @template param-pixel-size
#' @template param-max-d
#' @template param-min-density
#' @template param-min-votes
#' @section \code{LAS@data} Special Fields:
#'
#' Each point in the \code{LAS} object output represents a pixel center that is
#' \emph{possibly} also a stem cross-section center.
#'
#' The variables describing each point in the output are:
#'
#' \itemize{
#' \item \code{Intensity}: number of votes received by that point
#' \item \code{PointSourceID}: unique stem segment ID (among all trees)
#' \item \code{Keypoint_flag}: if \code{TRUE}, the point is the most likely circle center
#' of its stem segment (\code{PointSourceID})
#' \item \code{Radii}: approximate radius estimated by that point - always a multiple of the \code{pixel_size}
#' \item \code{TreeID}: unique tree ID of the point
#' \item \code{TreePosition}: if \code{TRUE}, the point represents the tree's position coordinate
#' }
#'
#' @template section-hough-transform
#' @section Tree Selection:
#'
#' An initial tree filter is used to select \emph{probable} trees in the input point
#' cloud. Parallel stacked layers, each one as thick as \code{hstep}, undergo the
#' circle search within the \code{hmin}/\code{hmax} limits. On every layer, pixels
#' above the \code{min_votes} criterion are clustered, forming
#' \emph{probability zones}. \emph{Probability zones} vertically aligned
#' on at least 3/4 of the stacked layers are assigned as \emph{tree occurrence regions}
#' and exported in the output map.
#'
#' @template reference-thesis
#' @template example-tree-map
#' @export
map.hough = function(min_h = 1, max_h = 3, h_step = 0.5, pixel_size = 0.025, max_d = 0.5, min_density = 0.1, min_votes = 3){

  if(max_h <= min_h)
    stop('max_h must be larger than min_h')

  params = list(
    h_step = h_step,
    pixel_size = pixel_size,
    max_d = max_d,
    min_density = min_density,
    min_votes = min_votes
  )

  for(i in names(params)){
    val = params[[i]]

    if(length(val) != 1)
      stop( i %>% paste('must be of length 1') )

    if(!is.numeric(val))
      stop( i %>% paste('must be Numeric') )

    if(val <= 0)
      stop( i %>% paste('must be positive') )
  }

  if(min_density > 1)
    stop('min_density must be between 0 and 1')

  func = function(las){

    rgz = las$Z %>% range

    if(max_h < rgz[1])
      stop('hmax is too low - below the point cloud')

    if(min_h > rgz[2])
      stop('hmin is too high - above the point cloud')

    map = stackMap(las %>% las2xyz, min_h, max_h, h_step, pixel_size, max_d/2, min_density, min_votes) %>%
      do.call(what=cbind) %>% as.data.table

    map$Intensity %<>% as.integer
    map$Keypoint_flag %<>% as.logical
    map$PointSourceID %<>% as.integer
    map$TreePosition %<>% as.logical
    map = suppressMessages(map %>% LAS %>% setHeaderTLS)

    return(map)
  }

  func %<>% setAttribute('tls_map_mtd')
  return(func)
}


#' Tree mapping algorithm: KNN point geometry
#' @description This function is meant to be used inside \code{\link{treeMap}}. It applies a KNN filter to select points with specific neighborhood features. For more details on geometry features, check out \code{\link{fastPointMetrics}}.
#' @template param-max-curvature
#' @template param-max-verticality
#' @param max_mean_dist \code{numeric} - maximum mean distance tolerated from a point to its nearest neighbors.
#' @template param-max-d
#' @template param-min_h-max_h
#' @template param-min-n
#' @details
#' Point metrics are calculated for every point. Points are then removed depending on their point
#' metrics parameters and clustered to represent individual tree regions.
#' Clusters are defined as a function of the expected maximum diameter. Any fields added to the
#' point cloud are described in \code{\link{fastPointMetrics}}.
#' @template section-eigen-decomposition
#' @export
map.eigen.knn = function(max_curvature = .1, max_verticality = 10, max_mean_dist = .1, max_d = .5, min_h = 1.5, max_h = 3, min_n = 100){

  params = list(
    max_curvature = max_curvature,
    max_verticality = max_verticality,
    max_mean_dist = max_mean_dist,
    max_d = max_d,
    min_h = min_h,
    max_h = max_h,
    min_n = min_n
  )

  for(i in names(params)){
    val = params[[i]]

    if(length(val) != 1)
      stop( i %>% paste('must be of length 1') )

    if(!is.numeric(val))
      stop( i %>% paste('must be Numeric') )

    if(val <= 0)
      stop( i %>% paste('must be positive') )
  }

  if(max_curvature > 1){
    stop('max_curvature must be a number between 0 and 1')
  }

  if(max_verticality > 180){
    stop('max_verticality must be a number between 0 and 180')
  }

  func = function(las){
    las = filter_poi(las, Classification != 2 & Z > min_h & Z < max_h)

    if(lidR::is.empty(las)){
      stop('no points found in the specified min_h/max_h range')
    }

    mtrlst = c('N', 'Curvature', 'Verticality', 'MeanDist')

    check_pt_metrics = mtrlst %>% sapply(function(x) hasField(las, x)) %>% as.logical %>% all
    if(!check_pt_metrics){
      message('Calculating knn fastPointMetrics')
      las = fastPointMetrics(las, ptm.knn(), mtrlst)
    }

    las = filter_poi(las, N > 3 & Curvature < max_curvature & abs(Verticality - 90) < max_verticality & MeanDist < max_mean_dist) %>%
      nnFilter(.1, 10)  %>% nnFilter(.25, 100)

    las@data$TreeID = 0
    maxdst = max_d*2
    i = 1
    while(any(las@data$TreeID == 0)){
      xy = las@data[TreeID == 0,.(X,Y)][1,] %>% as.double
      dst = las@data[,sqrt( (X-xy[1])^2 + (Y-xy[2])^2 )] %>% as.double
      las@data[TreeID == 0 & dst < maxdst]$TreeID = i
      i=i+1
    }

    hn = las@data[,.(H=max(Z) - min(Z), .N), by=TreeID]
    las = filter_poi(las, hn$N[TreeID] > min_n)

    las %<>% setAttribute('map_eigen')
    return(las)
  }

  func %<>% setAttribute('tls_map_mtd')
  return(func)
}


#' Tree mapping algorithm: Voxel geometry
#' @description This function is meant to be used inside \code{\link{treeMap}}. It applies a filter to select points belonging to voxels with specific features. For more details on geometry features, check out \code{\link{fastPointMetrics}}.
#' @template param-max-curvature
#' @template param-max-verticality
#' @param voxel_spacing \code{numeric} - voxel side length, in point cloud units.
#' @template param-max-d
#' @template param-min_h-max_h
#' @template param-min-n
#' @details
#' Point metrics are calculated for every voxel. Points are then removed depending on their voxel's metrics
#' metrics parameters and clustered to represent individual tree regions.
#' Clusters are defined as a function of the expected maximum diameter. Any fields added to the
#' point cloud are described in \code{\link{fastPointMetrics}}.
#' @template section-eigen-decomposition
#' @export
map.eigen.voxel = function(max_curvature = .15, max_verticality = 15, voxel_spacing = .1, max_d = .5, min_h = 1.5, max_h = 3, min_n = 100){

  params = list(
    max_curvature = max_curvature,
    max_verticality = max_verticality,
    voxel_spacing = voxel_spacing,
    max_d = max_d,
    min_h = min_h,
    max_h = max_h,
    min_n = min_n
  )

  for(i in names(params)){
    val = params[[i]]

    if(length(val) != 1)
      stop( i %>% paste('must be of length 1') )

    if(!is.numeric(val))
      stop( i %>% paste('must be Numeric') )

    if(val <= 0)
      stop( i %>% paste('must be positive') )
  }

  if(max_curvature > 1){
    stop('max_curvature must be a number between 0 and 1')
  }

  if(max_verticality > 180){
    stop('max_verticality must be a number between 0 and 180')
  }

  func = function(las){
    las = filter_poi(las, Classification != 2 & Z > min_h & Z < max_h)

    if(lidR::is.empty(las)){
      stop('no points found in the specified min_h/max_h range')
    }

    mtrlst = c('N', 'Curvature', 'Verticality')
    check_pt_metrics = c(mtrlst, 'VoxelID') %>% sapply(function(x) hasField(las, x)) %>% as.logical %>% all
    if(!check_pt_metrics){
      message('Calculating voxel fastPointMetrics')
      las = fastPointMetrics(las, ptm.voxel(voxel_spacing), mtrlst)
    }

    las = filter_poi(las, N > 3 & Curvature < max_curvature & abs(Verticality - 90) < max_verticality) %>%
      nnFilter(.1, 10)  %>% nnFilter(.25, 100)

    las@data$TreeID = 0
    maxdst = max_d*2
    i = 1
    while(any(las@data$TreeID == 0)){
      xy = las@data[TreeID == 0,.(X,Y)][1,] %>% as.double
      dst = las@data[,sqrt( (X-xy[1])^2 + (Y-xy[2])^2 )] %>% as.double
      las@data[TreeID == 0 & dst < maxdst]$TreeID = i
      i=i+1
    }

    hn = las@data[,.(H=max(Z) - min(Z), .N), by=TreeID]
    las = filter_poi(las, hn$N[TreeID] > min_n)

    las %<>% setAttribute('map_eigen')
    return(las)
  }

  func %<>% setAttribute('tls_map_mtd')
  return(func)
}


#' Tree mapping algorithm: pick trees manually
#' @description This function is meant to be used inside \code{\link{treeMap}}. It opens an interactive \code{rgl} plot where the user can specify tree locations by clicking.
#' @param map optional tree map to be manually updated.
#' @template param-min_h-max_h
#' @export
map.pick = function(map = NULL, min_h=1, max_h=5){

  if(min_h >= max_h){
    stop('max_h must be larger than min_h')
  }

  if(!is.null(map)){
    if(hasAttribute(map, 'tree_map_dt')){
      # map = map
    }else if(hasAttribute(map, 'tree_map')){
      map %<>% treeMap.positions(F)
    }else{
      stop('map is not a tree_map object: check ?treeMap')
    }
  }

  func = function(las){

    rgz = las$Z %>% range

    if(max_h < rgz[1])
      stop('hmax is too low - below the point cloud')

    if(min_h > rgz[2])
      stop('hmin is too high - above the point cloud')

    if(!is.null(min_h)){
      las = filter_poi(las, Z > min_h)
    }

    if(!is.null(max_h)){
      las = filter_poi(las, Z < max_h)
    }

    if(lidR::is.empty(las)){
      stop('no points found in the specified min_h/max_h range')
    }

    plot(las, size = .5, clear_artifacts=F)

    if(!is.null(map)){
      spheres3d(map$X, map$Y, median(las$Z), .33, color='white')
      text3d(map$X, map$Y, min(las$Z)-.5, map$TreeID, cex=1.5, col='yellow')
    }

    axes3d(col='white')
    pts = las@data %$% identify3d(X, Y, Z, tolerance = 50)

    ids = 1:length(pts)
    if(!is.null(map)){
      ids = ids + max(map$TreeID)
    }

    tmap = data.table(ids, las@data$X[pts], las@data$Y[pts]) # %>% toLAS(c('X','Y','TreeID'))
    names(tmap) = c('TreeID', 'X','Y')

    if(!is.null(map)){
      tmap = rbind(map, tmap)
    }

    tmap %<>% setAttribute('tree_map_dt')
    return(tmap)
  }

  func %<>% setAttribute('tls_map_mtd')
  return(func)
}
